react-native/react/engine/server.js

```diff
@@ -0,0 +1,82 @@
+'use strict'
+/* @flow */
+
+type Endpoint = (params: any) => (Promise<any> | any)
+
+type ServerEndpoints = { [key: string]: (ServerEndpoints | Endpoint) }
+type FlatServerEndpoints = { [key: string]: Endpoint }
+
+type Engine = any
+
+export default class Server {
+  startMethodName: string;
+  engine: Engine;
+  endpointsFn: (params: any, end: () => void) => FlatServerEndpoints;
+
+  constructor (engine: Engine, startMethodName: string, endMethodName: string, endpointMapFn: (params: any) => ServerEndpoints) {
+    this.engine = engine
+    this.startMethodName = startMethodName
+
+    this.endpointsFn = (params, end) => {
+      const endpoints = this.parseEndpoint(endpointMapFn(params))
+
+      // End the server when endMethodName is called
+      if (endMethodName) {
+        const originalEndMethod = endpoints[endMethodName]
+        endpoints[endMethodName] = (params) => {
+          end()
```

Comment:
order shouldn't matter, this is the last rpc call we'll get from the service, and all `end()` does is delete the session id from the incoming call map.